### DISCUSSION TITLE
Is django-guardian feature-complete?

### DISCUSSION THREAD

**Discussion by thclark** [django-guardian/django-guardian#821] [General]
OK. In this thread I want to talk about features, by which I mean significant new chunks of functionality.

So, we have a library that's 13 years old and has several thousands of stars on GitHub. Through that time, it's been used as one of the most core django libraries by presumably hundreds of thousands of developers.

If we filter the issues list by `enhancement` label we get [this list](https://github.com/django-guardian/django-guardian/issues?q=is%3Aissue+is%3Aopen+label%3Aenhancement), of which more than half the enhancement requests are more than a decade old at this point. Many of them, [like this one from 2011 that's still not made it](https://github.com/django-guardian/django-guardian/issues/23) might be extremely useful features, but might be better done as plugin libraries (think `django-guardian-whatever`) and referenced from the README here.

Despite none of those features **yet** making it in, guardian is extremely widely used and useful, suggesting these aren't 'killer' features. I also don't see people abandoning guardian because it doesn't have these things (typically people use either guardian, or an entirely different way of doing things). This suggests to me that they're mostly 'nice-to-haves', and further, actually looking at each issue, very few of them have much in the way of thumbs-up style support. Mostly there's a bit of a discussion by a very small number of interested parties, which also supports a conclusion of "better off in a third party plugin".

What are your thoughts? Is django-guardian feature-complete? Which ones of those features (or new ideas) do you think are key?

> Out of scope of this discussion: dependency updates, django releases compatibility, performance improvements/fixes, docs tweaks, updating CI/devops etc all of which (IMO) are basic requirements of ongoing maintenance.

**Comment by cmaggiulli** [comment]
I feel like we should prioritize fixes above enhancements, and any feature enhancements should be highly generic. For example in [810](https://github.com/django-guardian/django-guardian/issues/810) I proposed a new shortcut because we found that when implementing direct foreign key tables reassignment of permission objects is difficult.  

I think I would advise against that type of enhancement ( as contradictory as that may be ) - but maybe not if it's a bit more generic and it supports a larger feature ( like direct fk's ) that is certainly not feature complete as I describe in this slight twist [here](https://github.com/django-guardian/django-guardian/issues/810#issuecomment-2236390137/)

**Reply by thclark** [reply]
I couldn't agree more @cmaggiulli - I don't have the resource to be actually pursuing features myself, my intention is to do housekeeping on other people's PR and the repo itself, as well as fielding easier issues. It seems like you're a more experienced user, so I'll bow to your judgement on the precise balance of that.

On that particular issue, if I've understood correctly, you're proposing a model manager that would make the code you ended up with in the SO post easier to implement. I'm honestly not too sure whether that counts as a 'tweak' or a 'feature' so would suggest that you make a recommendation on whether to do it or not. Remember you can always take the vanilla way out and link to the SO answer from a new documentation section `Reassigning Permissions`!

**Reply by thclark** [reply]
As a general principle (assuming we don't go ahead with #741 which seems like it'll kill a large part of the user base, and bearing in mind @jmsmkns comments about api consistency below), I think it's a good guiding principle to say that "if we accept a PR it has to work for both FK and Generics, or be n/a for one of them (eg performance tweak specific to FKs only that doesn't affect generics)".

What do you think?

**Comment by jmsmkn** [comment]
Really glad that this library is getting attention again, thank you both for taking this on!

We have been using `django-guardian` in our application since July 2012 and honestly its been fantastic. It has also survived our external pentests. For us, we have not noticed anything which is seriously lacking, only sometimes the API feels a little incomplete or inconsistent.

For instance: `assign_perm` can take a list/queryset of groups/users and an object, `remove_perm` can only take one group/user and an object. Would be nice if all three arguments could be querysets (or lists) and guardian then works out how to most efficiently apply them. No big deal anyway.

We did run into a bit of a conceptual bug when integrating DRF which could have exposed some data. That requires setting the global permissions in order for the views to work. So we added them on top of the existing object permissions. `accept_global_perms` defaults to `False` in most places in `django-guardian`, _except_ in `get_objects_for_user` and `get_objects_for_group`! `get_objects_for_user` is then used in `PermissionsListMixin`, so then exposed all objects in the list views. Maybe a better default is to set this to `False` everywhere and have the user opt-in? But I understand, changing defaults is difficult when there are many existing people relying on this, but the consequence of that change would be that people wouldn't see objects they last had access to and could then opt in (and a safer default, IMO).

The only real problems we have faced have been performance, so of course we move to direct foreign key relations. Similar to https://github.com/django-guardian/django-guardian/issues/741#issuecomment-2171798229 we also hit performance problems and end up maintaining our own implementation of `filter_by_permission` (and everywhere that uses that). But, that's not too bad at all considering how big the application is now.

Thanks a lot for kicking this off again! As an active user I'm also happy to help out if you need some extra hands.

**Reply by thclark** [reply]
Thanks @jmsmkn we appreciate it! Very interesting comments.

I see four separate things here:

1. On assign/remove perm - I imagine a good first step would be if `remove_perm` accepted a list/queryset of groups/users. This at least would make them consistent in their API. I'd probably view that as a small enhancement that someone could contribute.

2. Accepting a list of objects and figuring out the most optimal way of applying everything, for both the generic and and foreign keyed solutions, would likely be a very significant piece of work that someone would have to have appetite for - and would likely first benefit from more foundational work around performance benchmarking, for example.
 
3. On the conceptual bug... top-level takeaway for me here is "if we were to make a breaking release (definitely not a given at this point), reconsidering the defaults for `accept_global_perms` might be a good change to include. If we aren't making a breaking release, perhaps a warning or further explanation in the documentation might be a good idea". I'd definitely want to understand why we ended up with that inconsistency in the first place.

4. If you've worked around @cmaggiulli 's performance problems with a custom `filter_by_permission` I don't know whether you'd be able to share that solution (if it's generic as opposed to containing business-specific logic) in an issue? Even if that constituted a breaking change and we decided not to actually merge it, it'd be great to pin it as a workaround for others to drop in.

Please do feel free to make issues for these, it's always beneficial to have these things tracked!

**Reply by jmsmkn** [reply]
Agree with all of that!

Our performance problem might not have been exactly the same but you can [find all of our adaptions here](https://github.com/comic/grand-challenge.org/blob/main/app/grandchallenge/core/guardian.py). The method in guardian was `get_objects_for_user` sorry, I just called our implementation `filter_by_permission`. If you would like a PR please let me know, but my thought is that they are too specific - it is only for DFKs, works best if you only have group permissions (which we almost always do) as then it becomes a straightforward reverse filter. However, if you need to consider user permissions too it uses a SQL Union operation. This is faster than the OR that is implemented in guardian but unfortunately you cannot further filter Union querysets in Django so it has to be the last operation, which isn't very user friendly.

I'll leave it up to you whether to create issues! I prefer this way of discussing things here then the owners of the repo get to decide on whether to take these things up, issues always seem a bit demanding (maybe that comes from my experience of leading an open source project...), especially when I don't think that anything is really wrong or missing at all.

**Reply by thclark** [reply]
Great, thanks @jmsmkn! It would work great for me too, as I only ever use group permissions. But I agree this is a bit too specific, and might just be an anchor for some people's implementations.

That said, perhaps it could be captured in an `Advanced` or `In The Wild` section of the docs, to highlight this as a technique without actually needing to do anything about the library itself.

My increasing sense, aligned with a view that there are generally workarounds out in the wild for most concerns people have, is that better documentation (and perhaps pinned issues) to lead people toward those workarounds is a much lighter-weight way of helping people overcome problems than rebooting guardian with new features.

This discussions group will help too.

**Reply by cmaggiulli** [reply]
> Thanks @jmsmkn we appreciate it! Very interesting comments.
> 
> I see four separate things here:
> 
> 4. If you've worked around @cmaggiulli 's performance problems with a custom `filter_by_permission` I don't know whether you'd be able to share that solution (if it's generic as opposed to containing business-specific logic) in an issue? Even if that constituted a breaking change and we decided not to actually merge it, it'd be great to pin it as a workaround for others to drop in.
> 
> Please do feel free to make issues for these, it's always beneficial to have these things tracked!

Will do.  I have a few things in the hopper unfortunately this month was a bit of crunch time at work so I havent gotten as much done as I would like.  We have an entirely generic middleware implementation of Guardian for DRF.  I think it was taken from the django guardian drf spinoff package and modified in some minor ways... we do not import that package as a dependency though.

Also @jmsmkn jogged my memory a bit.  There are concerns I have about Guardians implementation with DRF - especially if implemented with a json standard like JSONAPI or ODATA that allows for a "joining" of tables by way of an http request.  For example, using django-filters and django-rest-framework-json-api you can make calls such as

https://localhost:8000/v1/person?page[size]=300&filter[id]=123&include=address

which will return both the person data for person 123 and their referential address data.  This needs to be coded for if impl with drf and odata/jsonapi/other specs.  For us its not a problem because serendipitously all the joining thats happening is to tables that essentially "inherit" the permissions from the first table

Here is some relevant code ( I received approval from our CTO for this and other commits to this repo since we rely so heavily on this lib ):

```
#middleware/model_or_object_permissions.py
from guardian.shortcuts import get_objects_for_user
from rest_framework.filters import BaseFilterBackend
from rest_framework.permissions import DjangoObjectPermissions


class ModelOrObjectPermissions(DjangoObjectPermissions):
    perms_map = {
        "GET": ["%(app_label)s.view_%(model_name)s"],
        "OPTIONS": ["%(app_label)s.view_%(model_name)s"],
        "HEAD": ["%(app_label)s.view_%(model_name)s"],
        "POST": ["%(app_label)s.add_%(model_name)s"],
        "PUT": ["%(app_label)s.change_%(model_name)s"],
        "PATCH": ["%(app_label)s.change_%(model_name)s"],
        "DELETE": ["%(app_label)s.delete_%(model_name)s"],
    }

    def has_permission(self, request, view):
        setting = "force_check_model_permissions"
        force_check_model_permissions = getattr(view, setting, False)

        # if user is a superuser, grant all permissions
        if request.user.is_authenticated and request.user.is_superuser:
            return True

        if request.method == "POST" or force_check_model_permissions is True:
            return super().has_permission(request, view)
        else:
            # defer the check to `has_object_permission()`
            return True

    def has_object_permission(self, request, view, obj):
        setting = "_ignore_model_permissions"
        ignore_check_model_permissions = getattr(view, setting, False)

        if request.user.is_authenticated and request.user.is_superuser:
            return True

        if not ignore_check_model_permissions and super().has_permission(request, view):
            return True

        return super().has_object_permission(request, view, obj)


class ModelOrObjectPermissionsFilter(BaseFilterBackend):
    perm_format = "%(app_label)s.view_%(model_name)s"
    def filter_queryset(self, request, queryset, view):
        user = request.user
        model_cls = queryset.model
        kwargs = {
            "app_label": model_cls._meta.app_label,
            "model_name": model_cls._meta.model_name,
        }
        permission = self.perm_format % kwargs
        return get_objects_for_user(user, permission, queryset)
```

For delegation

```python
from sis_common.permissions import (
    ModelOrObjectPermissions,
    ModelOrObjectPermissionsFilter,
)


class DelegateView(object):
    """
    Delegates permission checking to a related object from another view.

    The main requirement for a view to use this class is that the model for
    the view must have a ForeignKey relation to a model specified by the
    required `delegate_to` class attribute. The `delegate_attr` attribute
    must also be set to specify the ForeignKey relationship.

    Intended for use with `DelegatePermissions` in `permission_classes`,
    as well as the `DelegatePermissionsFilter` as a `filter_backends`.

    Example usage:

        class DelegateCourseView(DelegateView):
            delegate_to = CourseViewSet
            delegate_attr = 'course'
            permission_classes = [
                And(
                    IsAuthenticated,
                    Or(Not(IsUsingTokenAuth), TokenMatchesOASRequirements),
                    DelegatePermissions
                ),
            ]
            filter_backends = (DelegatePermissionsFilter,)

    This will effectively delegate all permission checks and filtering to
    a related model for any views that use this as a mixin.
    """

    delegate_to = None
    delegate_attr = None

    class Delegator(object):
        def __init__(self, instance):
            self.instance = instance

    class ConfigurationError(Exception):
        """Raised if class is configured with missing attributes."""

    class MissingRelatedInstance(Exception):
        """Raised if related instance for `delegate_attr` does not exits."""

    class MissingSerializerRelatedData(Exception):
        """Raise if serializer `validated_data` is missing the `delegate_attr`."""

    def get_delegate_to(self):
        if self.delegate_to is None:
            raise self.ConfigurationError(
                "The `delegate_to` class attribute is missing."
            )
        return self.delegate_to

    def get_delegate_attr(self):
        if self.delegate_attr is None:
            raise self.ConfigurationError(
                "The `delegate_attr` class attribute is missing."
            )
        return self.delegate_attr

    def parse_delegate_attr(self, instance):
        delegate_attr_raw = self.get_delegate_attr()
        delegate_attrs = delegate_attr_raw.split(".")
        nested_relation = False

        if len(delegate_attrs) > 1:
            nested_relation = True
            delegate_attr_str = delegate_attrs.pop(0)
        else:
            delegate_attr_str = delegate_attr_raw

        try:
            delegate_field = getattr(instance, delegate_attr_str)
        except AttributeError:
            raise self.ConfigurationError(
                (
                    "The `delegate_attr` class attribute ({}) is not an "
                    "attribute of instance ({})."
                ).format(self.delegate_attr_raw, instance)
            )

        if not nested_relation:
            return delegate_field

        try:
            return getattr(delegate_field, "".join(delegate_attrs))
        except AttributeError:
            raise self.MissingRelatedInstance(
                f"Related instance for `{delegate_attr_raw}`."
            )

    def parse_serializer_data(self, validated_data):
        delegate_attr_raw = self.get_delegate_attr()
        delegate_attrs = delegate_attr_raw.split(".")
        nested_relation = False

        if len(delegate_attrs) > 1:
            nested_relation = True
            delegate_attr_str = delegate_attrs.pop(0)
        else:
            delegate_attr_str = delegate_attr_raw

        try:
            delegate_field = validated_data[delegate_attr_str]
        except KeyError:
            raise self.MissingSerializerRelatedData(
                (
                    "The `delegate_attr` key ({}) is missing from the serializer "
                    "`validated_data`."
                ).format(self.delegate_attr_raw)
            )

        if not nested_relation:
            return delegate_field

        try:
            return getattr(delegate_field, "".join(delegate_attrs))
        except AttributeError:
            raise self.MissingRelatedInstance(
                f"Related instance for `{delegate_attr_raw}`."
            )

    def perform_create(self, serializer):
        obj = self.parse_serializer_data(serializer.validated_data)
        self.check_object_permissions(self.request, self.Delegator(obj))
        super().perform_create(serializer)


class DelegatePermissions(ModelOrObjectPermissions):
    """
    Delegates permission checking to a related object from another view.

    This permission class can only be added to `permission_classes` for views
    of type `DelegateView`, in order to ensure `perform_create` is called with
    the necessary related object check.

    Note that users need the 'change_<delegate model>' permission instead of
    'add_<delegate model>' permission in order to add related data to a course
    (i.e. POST requests). This is because the data related to the delegate model
    is being changed. Override the `perms_map` if you prefer a different approach.
    """

    perms_map = {
        "GET": ["%(app_label)s.view_%(model_name)s"],
        "OPTIONS": ["%(app_label)s.view_%(model_name)s"],
        "HEAD": ["%(app_label)s.view_%(model_name)s"],
        "POST": ["%(app_label)s.change_%(model_name)s"],
        "PUT": ["%(app_label)s.change_%(model_name)s"],
        "PATCH": ["%(app_label)s.change_%(model_name)s"],
        "DELETE": ["%(app_label)s.delete_%(model_name)s"],
    }

    class InvalidViewError(Exception):
        """Raised if class is configured with invalid view."""

    def has_permission(self, request, view):
        if not isinstance(view, DelegateView):
            raise self.InvalidViewError(
                "Can only check permissions for views of type {}".format(
                    DelegateView.__name__
                )
            )

        if request.method == "POST":
            # DelegateView handles creation permissions in perform_create.
            return True

        return super().has_permission(request, view.delegate_to)

    def has_object_permission(self, request, view, obj):
        if not isinstance(view, DelegateView):
            raise self.InvalidViewError(
                "Can only check object permissions for views of type {}".format(
                    DelegateView.__name__
                )
            )

        if isinstance(obj, view.Delegator):
            # Called via check_object_permissions() in perform_create.
            return super().has_object_permission(
                request, view.delegate_to, obj.instance
            )

        d_obj = view.parse_delegate_attr(obj)
        return super().has_object_permission(request, view.delegate_to, d_obj)

    def _queryset(self, view):
        """
        The delegated object permissions must override this method
        of the `DjangoObjectPermissions` class to avoid this error:

            get_queryset() missing 1 required positional argument: 'self'

        This is because the `delegate_to` view passed to the delegate checker is
        a class and not an instance.
        """
        if getattr(view, "queryset", None) is None:
            raise ValueError(
                "Cannot apply {} on a view that does not set `.queryset`"
            ).format(self.__class__.__name__)
        return view.queryset


class DelegatePermissionsFilter(ModelOrObjectPermissionsFilter):
    """
    A filter backend that limits results to those where the requesting user
    has read object level permissions to a related model.

    Extends the ModelOrObjectPermissionsFilter, but filters by related models
    first to enforce related model permissions.

    Without this, since the default filter backend is ModelOrObjectPermissionsFilter,
    read permissions for the original model will be required or else a 404 error will
    be raised. To workaround, use this class as a filter backend instead.
    """

    def filter_queryset(self, request, queryset, view):
        related_queryset = super().filter_queryset(
            request, view.delegate_to.queryset, view.delegate_to
        )
        filter_args = {
            "{}__in".format(view.delegate_attr.replace(".", "__")): related_queryset
        }
        return queryset.filter(**filter_args)
```


Ive removed some comments to shrink size.  Im fully in support of enhancing documentation as opposed to adding features to the library... or, where applicable, making small changes which supports extensibility and then including code snippets in the readthedocs documentation to show some example implementations.  

We also have a port to get django guardian working with [django-elasticsearch-dsl-drf](https://pypi.org/project/django-elasticsearch-dsl-drf/) which I was planning on committing to their lib so we can just use that in our project

**Comment by BonaFideIT** [comment]
Hi,

as mentioned in https://github.com/django-guardian/django-guardian/pull/816#issuecomment-2323319279, we would like to see the maintenance of django-guardian continue.

From our point of view, one feature is particularly valuable, namely support for a custom group model in the same way as it works with the support of a custom user model.

There is an open PR https://github.com/django-guardian/django-guardian/pull/504 which suggestions a solution, but this results in 'missing' migration which changes the foreignkey relation between `GroupObjectPermissionBase` and the `auth.Group` model to the custom group model.

This migration is then created in the django-guardian migration directory and depends on the last migration of the app where the custom group model was placed. This is something you want to avoid, so we do not see this as a solution to the feature of using custom groups.

Here is an example how we would like to use django-guardian with a custom group model:

```python

class GuardianGroupMixin:
    def add_obj_perm(self, perm, obj):
        return GroupObjectPermission.objects.assign_perm(perm, self, obj)

    def del_obj_perm(self, perm, obj):
        return GroupObjectPermission.objects.remove_perm(perm, self, obj)


class CustomGroup(Group, GuardianGroupMixin):
    foo = models.CharField(max_length=120)
    ...


class CustomGroupPermissionsMixin(PermissionsMixin):
    groups = models.ManyToManyField(
        CustomGroup,
        verbose_name=_("groups"),
        blank=True,
        help_text=_(
            "The groups this user belongs to. A user will get all permissions "
            "granted to each of their groups."
        ),
        related_name="user_set",
        related_query_name="user",
    )

    class Meta(PermissionsMixin.Meta):
        abstract = True


class User(AbstractBaseUser, CustomGroupPermissionsMixin, GuardianUserMixin):
    email = models.EmailField(
        verbose_name=_("email address"), max_length=200, unique=True, db_index=True
    )
    ...

```

**Reply by BonaFideIT** [reply]
We added a PR (https://github.com/django-guardian/django-guardian/pull/825) to add some small changes to improve the custom group handling in django-guardian. Maybe that's enough, what do you think?
